### PR TITLE
Extension Compatiblity

### LINKED ISSUE #2356: Extension compatibility check
We need to implement a compatibility check for extensions.  An idea here is to mark the packages that we depend on that should be singletons, along with the associated semver.  We can then check the top level dependencies of an extension to see if they are compatible with this list, and mark it as invalid if the semver range does not overlap on any of these packages.  This check requires us to vendor the Python implementation of node semver that we had previously used.

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #2356.

Note: Populated the `singletonPackages` using the packages that export a token and `@phosphor/widgets`, which comprises the packages that are explicitly shared.

Sample output:

```
ValueError:
"@jupyterlab/python-tests@0.1.0" is not compatible with the current JupyterLab
Conflicting Dependencies:
Required	Actual	Package
^0.6.0  	^0.5.0	@jupyterlab/application
```

TODO:
- [x] Update the release script to add the explicit deps from the listed extensions
- [x] Use those deps along with singleton packages config to determine compat
- [x] Use the compat information in `list` - show as invalid, offer verbose mode with full report
- [x] Use the compat information when building - remove from templated `package.json`
- [x] Add tests

### CODE REVIEW

**Comment by blink1073** [comment]
Going to wait until #2409 is merged to finish this, since there will be a lot of overlap.

**Comment by ellisonbg** [comment]
Trying some things out with this branch:

First I ran:

`jupyter labextension uninstall @jupyterlab/help-extension`

I then saw this:

```bash
(python34) $ jupyter labextension list
[ListLabExtensionsApp] JupyterLab v0.23.2
[ListLabExtensionsApp] Known labextensions:
[ListLabExtensionsApp] 
Uninstalled core extensiosn:
[ListLabExtensionsApp]     @jupyterlab/help-extension
```

But when I tried to install that extension again I got a Python exception that ended with:

```python
ValueError: @jupyterlab/help-extension is not a valid JupyterLab extension
```

File: jupyterlab/commands.py [hunk 1]
```diff
@@ -210,160 +239,138 @@ def should_build(app_dir=None):
...
     if linked:
-        print('* Denotes linked packages')
+        logger.info('* Denotes linked extensions. Use `jupyter labextension listlinked` to see details')
 
     # Handle uninstalled and disabled core packages
     uninstalled_core = _get_uinstalled_core_extensions(app_dir)
     if uninstalled_core:
-        print('\nUninstalled core extensiosn:')
-        [print('    %s' % item) for item in sorted(uninstalled_core)]
+        logger.info('\nUninstalled core extensiosn:')
```
Comment (ellisonbg) [inline jupyterlab/commands.py]: extensiosn

**Comment by ellisonbg** [comment]
I tried to link/unlink the `mockextension` and `mockextension-incompat` and it all worked as expected.

**Comment by ellisonbg** [comment]
Tested latest changes locally and all is happy. Once the tests pass, let's merge!

**Comment by ellisonbg** [comment]
Come on Travis!

**Pull request merged**